src/main.cc

```diff
@@ -407,327 +928,183 @@ static void bro_new_handler()
+	char* tmp = script_signature_files;
+	char* s;
+	while ( (s = strsep(&tmp, " \t")) )
+		if ( *s )
+			rval.emplace_back(s);
 
-	if ( p )
-		add_to_name_list(p, ':', prefixes);
+	delete [] script_signature_files;
+	return rval;
+	}
 
-	string zeekygen_config;
+static std::string get_exe_path(const std::string& invocation)
+	{
+	if ( invocation.empty() )
+		return "";
 
-#ifdef USE_IDMEF
-	string libidmef_dtd_path = "idmef-message.dtd";
-#endif
+	if ( invocation[0] == '/' || invocation[0] == '~' )
+		// Absolute path
+		return invocation;
 
-	extern char* optarg;
-	extern int optind, opterr;
+	if ( invocation.find('/') != std::string::npos )
+		{
+		// Relative path
+		char cwd[PATH_MAX];
 
-	int long_optsind;
-	opterr = 0;
+		if ( ! getcwd(cwd, sizeof(cwd)) )
+			{
+			fprintf(stderr, "failed to get current directory: %s\n",
+			        strerror(errno));
+			exit(1);
+			}
 
-	char opts[256];
-	safe_strncpy(opts, "B:e:f:G:H:I:i:n:p:r:s:T:t:U:w:X:CFNPQSWabdhv",
-		     sizeof(opts));
+		return std::string(cwd) + "/" + invocation;
+		}
 
-#ifdef USE_PERFTOOLS_DEBUG
-	strncat(opts, "mM", 2);
-#endif
+	auto path = getenv("PATH");
 
-	int op;
-	while ( (op = getopt_long(bro_argc, bro_argv, opts, long_opts, &long_optsind)) != EOF )
-		switch ( op ) {
-		case 'a':
-			parse_only = true;
-			break;
+	if ( ! path )
+		return "";
 
-		case 'b':
-			bare_mode = true;
-			break;
+	return find_file(invocation, path);
+	}
 
-		case 'd':
-			fprintf(stderr, "Policy file debugging ON.\n");
-			g_policy_debug = true;
-			break;
+int main(int argc, char** argv)
+	{
+	ZEEK_LSAN_DISABLE();
+	std::set_new_handler(bro_new_handler);
 
-		case 'e':
-			command_line_policy = optarg;
-			break;
+	auto zeek_exe_path = get_exe_path(argv[0]);
 
-		case 'f':
-			user_pcap_filter = optarg;
-			break;
+	if ( zeek_exe_path.empty() )
+		{
+		fprintf(stderr, "failed to get path to executable '%s'", argv[0]);
+		exit(1);
+		}
 
-		case 'h':
-			usage(0);
-			break;
+	bro_argc = argc;
+	bro_argv = new char* [argc];
 
-		case 'i':
-			interfaces.push_back(optarg);
-			break;
+	for ( int i = 0; i < argc; i++ )
+		bro_argv[i] = copy_string(argv[i]);
 
-		case 'p':
-			prefixes.push_back(optarg);
-			break;
+	auto options = parse_cmdline(argc, argv);
 
-		case 'r':
-			read_files.push_back(optarg);
-			break;
+	if ( options.print_usage )
+		usage(argv[0], 0);
 
-		case 's':
-			rule_files.push_back(optarg);
-			break;
+	if ( options.print_version )
+		{
+		fprintf(stdout, "%s version %s\n", argv[0], zeek_version());
+		exit(0);
+		}
 
-		case 't':
-			g_trace_state.SetTraceFile(optarg);
-			g_trace_state.TraceOn();
-			break;
+	if ( options.run_unit_tests )
+		{
+		doctest::Context context;
+		auto dargs = to_cargs(options.doctest_args);
+		context.applyCommandLine(dargs.size(), dargs.data());
+		ZEEK_LSAN_ENABLE();
+		return context.run();
+		}
 
-		case 'v':
-			fprintf(stdout, "%s version %s\n", prog, zeek_version());
-			exit(0);
-			break;
+	pid_t stem_pid = 0;
```

Comment:
Add some comments here what's going on with the stem process & env variable, not clear without also reading the supervisor code.

Also, move more functionality into the Supervisor?

Reply:
Sure, agree it's better organizing all that behind a single `Supervisor::CreateStem()` call.